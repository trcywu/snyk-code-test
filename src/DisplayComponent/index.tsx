import React, { FC } from "react";
import { makeStyles } from "@material-ui/core/styles";
import { TreeView, TreeItem } from "@material-ui/lab";
import ExpandMoreIcon from "@material-ui/icons/ExpandMore";
import ChevronRightIcon from "@material-ui/icons/ChevronRight";
import { Container } from "@material-ui/core";

const useStyles = makeStyles({
  container: {
    marginTop: 24
  },
  root: {
    height: 240,
    flexGrow: 1,
    maxWidth: 400
  }
});

export interface DataResponse {
  name: string;
  version: string;
  dependencies: Record<string, string>;
  devDependencies: Record<string, string>;
  peerDependencies: Record<string, string>;
}

interface Props {
  displayList: DataResponse;
}

const DisplayComponent: FC<Props> = ({ displayList }) => {
  const classes = useStyles();
  return displayList && displayList.name ? (
    <Container className={classes.container}>
      <TreeView
        className={classes.root}
        defaultCollapseIcon={<ExpandMoreIcon />}
        defaultExpandIcon={<ChevronRightIcon />}
      >
        <TreeItem nodeId="name" label={displayList.name}>
          <TreeItem nodeId="version" label={`v${displayList.version}`} />

          <TreeItem nodeId="dependencies" label="dependencies">
            {displayList.dependencies ? (
              Object.entries(displayList.dependencies).map(
                (item: Array<string>) => {
                  return (
                    <TreeItem
                      key={`${item[0]}`}
                      nodeId={`${item[0]}`}
                      label={`${item[0]} @${item[1]}`}
                    />
                  );
                }
              )
            ) : (
              <TreeItem nodeId="noDependencies" label="none" />
            )}
          </TreeItem>

          <TreeItem nodeId="devDependencies" label="devDependencies">
            {displayList.devDependencies ? (
              Object.entries(displayList.devDependencies).map(
                (item: Array<string>) => {
                  return (
                    <TreeItem
                      key={`${item[0]}`}
                      nodeId={`${item[0]}`}
                      label={`${item[0]} @${item[1]}`}
                    />
                  );
                }
              )
            ) : (
              <TreeItem nodeId="noDevDependencies" label="none" />
            )}
          </TreeItem>

          <TreeItem nodeId="peerDependencies" label="peerDependencies">
            {displayList.peerDependencies ? (
              Object.entries(displayList.peerDependencies).map(
                (item: Array<string>) => {
                  return (
                    <TreeItem
                      key={`${item[0]}`}
                      nodeId={`${item[0]}`}
                      label={`${item[0]} @${item[1]}`}
                    />
                  );
                }
              )
            ) : (
              <TreeItem nodeId="noPeerDependencies" label="none" />
            )}
          </TreeItem>
        </TreeItem>
      </TreeView>
    </Container>
  ) : null;
};

export default DisplayComponent;
