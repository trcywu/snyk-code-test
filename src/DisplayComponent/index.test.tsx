import React from "react";
import renderer from "react-test-renderer";
import DisplayComponent from ".";
import { DataResponse } from "../DisplayComponent";

const createNodeMock = (element: React.ReactElement<any, string>) => {
  if (element.type === "div") {
    return {
      textContent: ""
    };
  }
  return null;
};

const dataResponse: DataResponse = {
  name: "express",
  version: "1.2.3",
  dependencies: {
    test1: "~1.3.7"
  },
  devDependencies: {
    test2: "0.8.2"
  },
  peerDependencies: {
    test3: "4.2.1"
  }
};

describe("<DisplayComponent />", () => {
  it("renders DisplayComponent with displayList", async () => {
    let wrapper = renderer.create(
      <DisplayComponent displayList={dataResponse} />,
      { createNodeMock }
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it("renders DisplayComponent with empty displayList", async () => {
    let wrapper = renderer.create(
      <DisplayComponent displayList={{} as DataResponse} />,
      { createNodeMock }
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
