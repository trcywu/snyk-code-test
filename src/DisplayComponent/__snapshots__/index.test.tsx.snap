// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DisplayComponent /> renders DisplayComponent with displayList 1`] = `
<div
  className="MuiContainer-root makeStyles-container-1 MuiContainer-maxWidthLg"
>
  <ul
    aria-multiselectable={false}
    className="MuiTreeView-root makeStyles-root-2"
    role="tree"
  >
    <li
      aria-expanded={false}
      aria-selected={false}
      className="MuiTreeItem-root"
      onFocus={[Function]}
      onKeyDown={[Function]}
      role="treeitem"
      tabIndex={-1}
    >
      <div
        className="MuiTreeItem-content"
        onClick={[Function]}
        onMouseDown={[Function]}
      >
        <div
          className="MuiTreeItem-iconContainer"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
            />
          </svg>
        </div>
        <div
          className="MuiTypography-root MuiTreeItem-label MuiTypography-body1"
        >
          express
        </div>
      </div>
    </li>
  </ul>
</div>
`;

exports[`<DisplayComponent /> renders DisplayComponent with empty displayList 1`] = `null`;
